mypy `check` ignores files that don't end with `.py`

**Describe the bug**

Consider the following BUILD file:
```python
python_sources(
    name="python-scripts",
    sources=[
        "python-script-without-extension",
    ],
)
```

The command `pants check ::` will not lint `python-script-without-extension` with mypy.

However `format` will work. So I assume that this bug has to do with `check` and not with the python backend?

**Pants version**
- `2.22.0`

Sorry for the trouble. Standalone reproducer:

```shell
cd $(mktemp -d)

cat > pants.toml <<EOF
[GLOBAL]
pants_version = "2.22.0"

backend_packages = [
  "pants.backend.python",
  "pants.backend.python.typecheck.mypy",
]

[python]
interpreter_constraints = ["CPython==3.10.*"]
EOF

cat > BUILD <<EOF
python_sources(
    name="python-scripts",
    sources=["without-extension", "with-extension.py"],
)
EOF

echo '1 + "no .py"' > without-extension
echo '1 + "has .py"' > with-extension.py

pants check :: # BUG: only with-extension.py has an error
```

Running that like `bash script.sh` gives this output:

```
with-extension.py:1: error: Unsupported operand types for + ("int" and "str")  [operator]
Found 1 error in 1 file (checked 1 source file)
```

But there should be 2 errors, across 2 files, including `without-extension`.

Running with `pants --keep-sandboxes=on_failure check ::` to explore the sandbox finds:

- the sandbox contains both both `with-extension.py` and `without-extension` ✅ 
- `__run.sh` invokes to `__mypy_runner.sh` ...
- `__mypy_runner.sh` has an invocation like `./mypy.pex_pex_shim.sh --python-executable=./requirements_venv.pex_bin_python_shim.sh --python-version=3.10 --skip-cache-mtime-check --sqlite-cache --cache-dir .tmp_cache/mypy_cache @__files.txt`
- `__files.txt` contains just `with-extension.py` ❌ 

Searching the pants repo for `__files.txt` finds this bit of code in the mypy backend:

https://github.com/pantsbuild/pants/blob/c3097a2fb61d2a1733d7a01301940e23f74d6def/src/python/pants/backend/python/typecheck/mypy/rules.py#L197-L202

And indeed `determine_python_files` looks at extensions:

https://github.com/huonw/pants/blob/c3097a2fb61d2a1733d7a01301940e23f74d6def/src/python/pants/backend/python/typecheck/mypy/rules.py#L114-L130

So, that somehow needs to be adjusted to handle files without an extension.

I think there should be an `else` branch that just adds the file name unconditionally. @aviau would you like to contribute this fix?

--- a/docs/notes/2.24.x.md
+++ b/docs/notes/2.24.x.md
@@ -82,6 +82,7 @@ Version Updates:
 
 A new experimental [Python Provider](https://www.pantsbuild.org/blog/2023/03/31/two-hermetic-pythons) using [Python Build Standalone](https://gregoryszorc.com/docs/python-build-standalone/main/) is available as `pants.backend.python.providers.experimental.python_build_standalone`.  This joins the existing [pyenv provider](https://www.pantsbuild.org/stable/reference/subsystems/pyenv-python-provider) as a way for Pants to take care of providing an appropriate Python.
 
+Mypy will now typecheck previously-ignored python sources without a `.py` or `.pyi` extension.
 
 #### S3
 

--- a/src/python/pants/backend/python/typecheck/mypy/rules.py
+++ b/src/python/pants/backend/python/typecheck/mypy/rules.py
@@ -127,6 +127,9 @@ def determine_python_files(files: Iterable[str]) -> Tuple[str, ...]:
             pyi_file = f + "i"
             if pyi_file not in result:
                 result.add(f)
+        else:
+            result.add(f)
+
     return tuple(result)
 
 

--- a/src/python/pants/backend/python/typecheck/mypy/rules_integration_test.py
+++ b/src/python/pants/backend/python/typecheck/mypy/rules_integration_test.py
@@ -694,7 +694,7 @@ def test_determine_python_files() -> None:
     assert determine_python_files(["f.pyi"]) == ("f.pyi",)
     assert determine_python_files(["f.py", "f.pyi"]) == ("f.pyi",)
     assert determine_python_files(["f.pyi", "f.py"]) == ("f.pyi",)
-    assert determine_python_files(["f.json"]) == ()
+    assert determine_python_files(["script-without-extension"]) == ("script-without-extension",)
 
 
 def test_colors_and_formatting(rule_runner: PythonRuleRunner) -> None:

